### PR TITLE
feat: Call Analyzer -Breakout Move Metrics

### PR DESCRIPTION
**Pull request by lhy2143**
# COMPLETES 
https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-404837

## This pull request addresses

send metrics when moving to a breakout.

## by making the following changes

add code to send metrics when moving to a breakout.
set the meeting object as a parent for breakouts.
update README.MD.
add unit test.

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

< ENUMERATE TESTS PERFORMED, WHETHER MANUAL OR AUTOMATED >

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/@webex/plugin-meetings/src/breakouts/breakout.ts [hunk 1]
```diff
@@ -56,14 +58,38 @@ const Breakout = WebexPlugin.extend({
...
+    Metrics.postEvent({
+      event: eventType.MOVE_TO_BREAKOUT,
+      meetingId: activeMeetingId,
+      data: {
+        breakoutMoveId,
+        breakoutSessionId: this.sessionId,
+        breakoutGroupId: this.groupId,
+      },
+    });
+    const result = this.request({
```
Comment (arun3528) [inline packages/@webex/plugin-meetings/src/breakouts/breakout.ts:72]: i would just use await here , more readable

File: packages/@webex/plugin-meetings/src/breakouts/breakout.ts [hunk 2]
```diff
@@ -56,14 +58,38 @@ const Breakout = WebexPlugin.extend({
    * @returns {Promise}
    */
   join() {
-    return this.request({
+    const breakoutMoveId = uuid.v4();
+    const activeMeetingId = this.collection.parent.meeting.deref().id;
```
Comment (arun3528) [inline packages/@webex/plugin-meetings/src/breakouts/breakout.ts:62]: its a hacky way to get the meeting id here cant we save the meeting id when the breakout session gets created

File: packages/@webex/plugin-meetings/src/breakouts/index.ts [hunk 3]
```diff
@@ -255,6 +256,7 @@ const Breakouts = WebexPlugin.extend({
     forEach(breakouts, (breakout: typeof Breakout) => {
       // eslint-disable-next-line no-param-reassign
       breakout.url = this.url;
+      breakout.parent = this;
```
Comment (arun3528) [inline packages/@webex/plugin-meetings/src/breakouts/index.ts:259]: i feel this is like a circular dependency

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 4]
```diff
@@ -609,7 +609,7 @@ export default class Meeting extends StatelessWebexPlugin {
      * @memberof Meeting
      */
     // @ts-ignore
-    this.breakouts = new Breakouts({}, {parent: this.webex});
+    this.breakouts = new Breakouts({meeting: new WeakRef(this)}, {parent: this.webex});
```
[the remark this answers is not in the document]
Comment (lhy2143) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:612]: okay. the param must be an object so I pass {id: this.id}

File: packages/@webex/plugin-meetings/src/breakouts/index.ts (hunk 3 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts:259]: Issue: This isn't necessary. Each breakout is already in a collection and the collection's parent is the breakouts object where you store the meetingId. To access the meetingId from the breakout:
```ts
this.collection.parent.meetingId
```

File: packages/@webex/plugin-meetings/src/breakouts/breakout.ts [hunk 5]
```diff
@@ -55,15 +57,37 @@ const Breakout = WebexPlugin.extend({
    * Joins the breakout session
    * @returns {Promise}
    */
-  join() {
-    return this.request({
+  async join() {
+    const breakoutMoveId = uuid.v4();
+    const {meetingId} = this.parent;
```
Comment (Coread) [inline packages/@webex/plugin-meetings/src/breakouts/breakout.ts:62]: As mentioned in the other comment to access the breakouts object from a breakout you can write something like
```ts
const {meetingId} = this.collection.parent
```

**Review by Coread — CHANGES_REQUESTED** [review]
please see comment

File: packages/@webex/plugin-meetings/test/unit/spec/breakouts/breakout.ts [hunk 6]
```diff
@@ -45,6 +51,11 @@ describe('plugin-meetings', () => {
 
         assert.equal(result, 'REQUEST_RETURN_VALUE');
       });
+      it('send metrics as expected', async () => {
+        Metrics.postEvent = sinon.stub();
+        await breakout.join();
+        assert.calledTwice(Metrics.postEvent);
```
Comment (Coread) [inline packages/@webex/plugin-meetings/test/unit/spec/breakouts/breakout.ts:57]: issue: It is important to check the arguments of the calls to Metrics.postEvent.

**Review by Coread — CHANGES_REQUESTED** [review]
Please see comment about checking the arguments of function calls in the unit test

**Review by Coread — APPROVED** [review]

**Pull request merged**